### PR TITLE
ENH: Implement `np.floating.as_integer_ratio`

### LINKED ISSUE #9969: ENH: Implement `np.floating.as_integer_ratio`
To match the builtin `float.as_integer_ratio`

This would allow conversion from `longdouble` to `Decimal`, and to `Fraction` if https://bugs.python.org/issue28716 is resolved

### PR DESCRIPTION
**Pull request by eric-wieser**
The original incentive here was to port tests from CPython for #9963

Fixes #9969

### CODE REVIEW

File: numpy/core/tests/test_scalar_methods.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+        for f, ratio in [
+                (0.875, (7, 8)),
+                (-0.875, (-7, 8)),
+                (0.0, (0, 1)),
+                (11.5, (23, 2)),
+            ]:
+            assert_equal(ftype(f).as_integer_ratio(), ratio)
+
+        for i in range(10000):
```
Comment (mhvk) [inline numpy/core/tests/test_scalar_methods.py]: I realize the tests are from cpython, but I still dislike throwing random numbers at tests - it just means that things become non-reproducible. Why not explicitly pick the smallest and largest representable number? And do `2.1` for all the types.

**Review by mhvk — COMMENTED** [review]
Beyond the comment about the test, I think this needs an entry in the release notes - I presume this is mostly there for compatibility with cpython float?

File: numpy/core/tests/test_scalar_methods.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: Do you want me to fix the random seed?

Note the result when applied to 2.1 is different for each type (and platform dependent for float128)

Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: Do you want me to fix the random seed?

Note the result when applied to 2.1 is different for each type (and platform dependent for float128)

Comment (mhvk) [reply numpy/core/tests/test_scalar_methods.py]: I'd just prefer not to use random numbers at all, and not 1000 of them either (needless increase of test time), i.e., just loop over some 10 examples chosen with some care (just testing as you do now that the result is consistent).

And, yes, you're right, for 2.1 the results would be different -- it's fine not to include 128 for that, given the platform dependence.

Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: I think my ability to choose relevant corner cases is very limited, so I'd prefer to keep the randomness. I've reduced the number of iterations from 10000 to 1000, seeded the randomstate, and added explicit tests for 2.1.

Note that the largest representable number is not testable, as its denominator may be larger than itself. The relevant tests are the ones that use lots of mantissa bits.

**Comment by eric-wieser** [comment]
Updated with docstring and release notes

File: numpy/core/tests/test_scalar_methods.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/tests/test_scalar_methods.py]: OK, this was not meant to be a big deal.

**Comment by njsmith** [comment]
I assume this is inspired by the debate raging in python core:

* https://bugs.python.org/issue26680
* https://mail.python.org/pipermail/python-dev/2018-March/152358.html
* https://bugs.python.org/issue33073

The argument there is actually about adding `as_integer_ratio` to `int`, which might suggest doing the same for the `np.integer` types. Of course our ints are less duck-compatible with floats than Python's ints, because if you try to treat say a `np.int8` as if it were a `float` then you're going to have a bad time with wraparound.

The whole topic seems very unimportant to me, because who even cares about `as_integer_ratio` on floats *or* ints? It seems like parlor trick. If you want to find out the exact value of a float there are more useful ways. But OTOH I guess this isn't necessarily an argument against it either.

I'm wincing at adding more incompatibilities between scalars and 0d arrays. OTOH I don't know that eliminating those is actually realistic, and we already have this on `float64`, and this is unlikely to add major new obstacles since it seems unlikely anyone will use it much.

It's also unusual and surprising that this returns Python ints rather than numpy ints, but OTOH I guess this is necessary because the returned values might not fit into an `int64`.

So kinda mixed feelings about this overall. I guess my overall impression is an emphatic meh?

Am I missing important use cases?

**Comment by njsmith** [comment]
Oh, on further examination it looks like that started out as a debate about the `float.is_integer` method, which then dragged in `as_integer_ratio` as it went. Sorry, this is confusing...

`is_integer` is much more of an attractive nuisance than `as_integer_ratio`, since people might actually think they want it, but they usually don't (because of the usual floating point rounding issues).

**Comment by eric-wieser** [comment]
> The argument there is actually about adding `as_integer_ratio` to `int`, which might suggest doing the same for the `np.integer` types

I saw something like that elsewhere, but decided to wait to see what CPython does first. As they point out, most of the value is in consistency - and what we do right now is consistent with CPython!

> If you want to find out the exact value of a float there are more useful ways. 

I don't think this is true for `np.longdouble`. You could use `Decimal(str(f))`, but that loses precision.



>  I guess this is necessary because the returned values might not fit into an int64.

Yep, otherwise we may as well implement `longdouble.as_integer_ratio` as `float(ld).as_integer_ratio()` which drops precision, defeating the point of the function

> I'm wincing at adding more incompatibilities between scalars and 0d arrays

I'm hoping that eventually we'll solve the dtype-specific-methods-on-arrays problem

**Comment by njsmith** [comment]
>> If you want to find out the exact value of a float there are more useful ways.

> I don't think this is true for `np.longdouble`. You could use `Decimal(str(f))`, but that loses precision.

Wait, but wasn't that the whole point of the Dragon4 work, that `str(f)` shouldn't lose precision? Does it not work for `longdouble`? This seems like a bug to fix regardless.

> I'm hoping that eventually we'll solve the dtype-specific-methods-on-arrays problem

I don't see how any solution is really possible, since if we allow arbitrary methods to be added to arrays then it means we can never add any new methods to arrays ourselves without potentially breaking backcompat :-/.

**Comment by seberg** [comment]
Sorry, I realized this is probably just spitting out noise, but since I had started writing, I did not want to delete it again ;).

My (probably naive) gut feeling on dtype specific methods is to allow something like an `elementwise` property which could give access to ufuncs (or at least unary ufuncs) and could mimic such methods. But then I also am of the opinion that scalars are a good idea, and I seem to be in the minority there :) (not in the sense of code duplication of course, but because of things like mutability and because I actually believe that in almost all cases it would be "obvious" if the result will be scalar or array).

The main "oddity" about compatibility is in my opinion not the 0D arrays do not behave like scalars, but the part that scalars have to pretend they are 0D arrays because we convert 0D arrays to scalars for good reason almost everywhere. It seems to me that this incompatibility is the other way around though, so I do not really see much of an issue ;).

I would almost like something like:
```
class MyFloat(np.float64):
    add_method_from_ufunc(ufunc_implementation, 'possibly_method_name')

    # Or for convenience, define it all in python (slow, but....):
    @as_ufunc(out_dtype=object)
    def as_integer_ratio(self):
        # Numpy can wrap it into a ufunc:
        return integer_ratio
```
Now, will people be able to block our methods? Sure, but only if they subclass our types, and then we could possibly even implement a warning that a new dtype method is being/will be shadowed.

**Comment by eric-wieser** [comment]
> Wait, but wasn't that the whole point of the Dragon4 work, that `str(f)` shouldn't lose precision? 

No, I think the point was that `ftype(str(f))` shouldn't lose precision. Dragon4 produces a unique representation for each float, but I don't think it produces an exact decimal representation.

For example:
```python
>>> from fractions import Fraction
>>> f16 = np.float16(2.1)

>>> f_bad = Fraction(str(f16))
Fraction(21, 10)
>>> f_bad == f16
False

>>> f_good = Fraction(*f16.as_integer_ratio())
Fraction(1075, 512)
>>> f_good == f16
True
```

**Comment by eric-wieser** [comment]
One of the incentives behind this PR was to be able to write better tests for `longdouble` types - in particular, the cpython tests for `math.remainder` fall back on `.as_integer_ratio()`, so it would nice to be able to duplicate them to test `remainder_ieee` #9963

**Comment by eric-wieser** [comment]
Any more thoughts on this? I don't think that thinking about dtype-specific array methods is relevant here - we already have `np.float64.as_integer_ratio`, adding `np.floatXX.as_integer_ratio` seems like a non-contentious extension.

The only argument I can see against this is maintenance cost - and I think that's paid for by a better ability to debug precision issues with `longdouble`.

**Comment by mhvk** [comment]
I think we might as well have it.

**Comment by eric-wieser** [comment]
@charris, any opinions?

CI failure is garbage here

**Comment by mattip** [comment]
Should/Did this hit the mailing list? Seems like a simple case of adapting to CPython's float type. In any case it needs updating for 1.16

**Comment by eric-wieser** [comment]
Conflicts resolved

**Comment by eric-wieser** [comment]
Python is now getting `int.as_integer_ratio()` too (python/cpython#8750), so it seems like the decision is that all numeric types should have this.

**Comment by eric-wieser** [comment]
Conflicts resolved

**Comment by eric-wieser** [comment]
Guess I now need to update this for 1.17...

**Comment by mhvk** [comment]
I'd be quite happy to merge this once rebased.

**Comment by eric-wieser** [comment]
Rebased, let's see if tests still pass

File: numpy/core/tests/test_scalar_methods.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    run_module_suite,
+    assert_equal, assert_almost_equal, assert_raises, assert_warns,
+    dec
+)
+
+float_types = [np.half, np.single, np.double, np.longdouble]
+
+def test_float_as_integer_ratio():
+    # derived from the cpython test "test_floatasratio"
+    for ftype in float_types:
```
Comment (mhvk) [inline numpy/core/tests/test_scalar_methods.py]: Could now use `@pytest.mark.parametrize` but no big deal.

File: numpy/core/tests/test_scalar_methods.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                (0.875, (7, 8)),
+                (-0.875, (-7, 8)),
+                (0.0, (0, 1)),
+                (11.5, (23, 2)),
+            ]:
+            assert_equal(ftype(f).as_integer_ratio(), ratio)
+
+        rstate = np.random.RandomState(0)
+        fi = np.finfo(ftype)
+        for i in range(1000):
```
Comment (mhvk) [inline numpy/core/tests/test_scalar_methods.py]: I don't see the point of repeating the same 1000 numbers every time, especially with a `try/except` - I think it is much better to use actual examples, say `np.pi`, `sqrt(2)`, `1/9`, including ones where the `try/except` would fail (presumably very large or small exponent).

**Review by mhvk — COMMENTED** [review]
Implementation looks all OK. For the tests, I'd prefer replacing the 1000 always-the-same random samples with a few well-chosen ones.

File: numpy/core/tests/test_scalar_methods.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    assert_equal(R(-4404019, 2097152),
+                 R(*np.single(-2.1).as_integer_ratio()))
+    assert_equal(R(4728779608739021, 2251799813685248),
+                 R(*np.double(2.1).as_integer_ratio()))
+    assert_equal(R(-4728779608739021, 2251799813685248),
+                 R(*np.double(-2.1).as_integer_ratio()))
+    # longdouble is platform depedent
+
+
+if __name__ == "__main__":
```
Comment (tylerjereddy) [inline numpy/core/tests/test_scalar_methods.py]: This looks like a `nose` thing we don't need anymore?

File: numpy/core/tests/test_scalar_methods.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                     R(*ftype(0.5).as_integer_ratio()))
+        assert_equal(R(-2100, 1),
+                     R(*ftype(-2100.0).as_integer_ratio()))
+
+        assert_raises(OverflowError, ftype('inf').as_integer_ratio)
+        assert_raises(OverflowError, ftype('-inf').as_integer_ratio)
+        assert_raises(ValueError, ftype('nan').as_integer_ratio)
+
+
+    assert_equal(R(1075, 512),
```
Comment (tylerjereddy) [inline numpy/core/tests/test_scalar_methods.py]: The indentation level on this block of asserts suggests they can be broken off into a separate test that is not related to the parametrization / iteration over types / values, right?

File: numpy/core/tests/test_scalar_methods.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                df = np.longdouble(str(d))
+            except (OverflowError, RuntimeWarning):
+                # the values may not fit in any float type
+                continue
+
+            assert_equal(
+                dn / df, f,
+                "{}/{} (dtype={})".format(n, d, ftype.__name__))
+
+        R = fractions.Fraction
```
Comment (tylerjereddy) [inline numpy/core/tests/test_scalar_methods.py]: The nesting makes the tests below a little hard to follow -- only need `ftype` iteration / parametrization here I think?

File: numpy/core/tests/test_scalar_methods.py (hunk 3 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: Sounds like a good fit for [`hypothesis`](https://github.com/HypothesisWorks/hypothesis) to generate some cases for us.

File: numpy/core/tests/test_scalar_methods.py (hunk 2 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: Working on this locally a bit, I ended up splitting into a few different tests, with a really small one that looks like this to cut down on the nesting:

```python
 16 @pytest.mark.parametrize("ftype", [
 17     np.half, np.single, np.double, np.longdouble])
 18 @pytest.mark.parametrize("f, ratio", [
 19     (0.875, (7, 8)),
 20     (-0.875, (-7, 8)),
 21     (0.0, (0, 1)),
 22     (11.5, (23, 2)),
 23     ])
 24 def test_float_as_integer_ratio(ftype, f, ratio):
 25     # derived from the cpython test "test_floatasratio"
 26     assert_equal(ftype(f).as_integer_ratio(), ratio)
```

**Review by tylerjereddy — COMMENTED** [review]
I started refactoring the test a little locally based on Marten's feedback. Maybe I'll push a separate commit for Eric to revise.

**Comment by tylerjereddy** [comment]
I pushed in a draft commit that breaks the huge / nested loop test into smaller tests for clarity with respect to reporting failures and having less nesting in a single test. Not sure if Eric will like that but can always just gut the commit.

I didn't try to tackle the case generation yet though.

**Comment by eric-wieser** [comment]
I think a lot of the test structure was designed to resemble the upstream tests this emulates. I'll have to check how much I modified them

**Comment by eric-wieser** [comment]
This is the test I copied from:
```python
    def test_floatasratio(self):
        for f, ratio in [
                (0.875, (7, 8)),
                (-0.875, (-7, 8)),
                (0.0, (0, 1)),
                (11.5, (23, 2)),
            ]:
            self.assertEqual(f.as_integer_ratio(), ratio)

        for i in range(10000):
            f = random.random()
            f *= 10 ** random.randint(-100, 100)
            n, d = f.as_integer_ratio()
            self.assertEqual(float(n).__truediv__(d), f)

        R = fractions.Fraction
        self.assertEqual(R(0, 1),
                         R(*float(0.0).as_integer_ratio()))
        self.assertEqual(R(5, 2),
                         R(*float(2.5).as_integer_ratio()))
        self.assertEqual(R(1, 2),
                         R(*float(0.5).as_integer_ratio()))
        self.assertEqual(R(4728779608739021, 2251799813685248),
                         R(*float(2.1).as_integer_ratio()))
        self.assertEqual(R(-4728779608739021, 2251799813685248),
                         R(*float(-2.1).as_integer_ratio()))
        self.assertEqual(R(-2100, 1),
                         R(*float(-2100.0).as_integer_ratio()))

        self.assertRaises(OverflowError, float('inf').as_integer_ratio)
        self.assertRaises(OverflowError, float('-inf').as_integer_ratio)
        self.assertRaises(ValueError, float('nan').as_integer_ratio)
```

**Comment by mhvk** [comment]
Hmm, yes, I think I complained about the random number test before :-; which is probably why you have only 1000, not 10000. Still do not see any point in it.

**Comment by eric-wieser** [comment]
> I pushed in a draft commit that breaks the huge / nested loop test into smaller tests

If we're breaking up the test, it would make sense to me to create a `class TestAsIntegerRatio` to group them together.

File: numpy/core/tests/test_scalar_methods.py (hunk 3 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: I've replaced the 1000 iterations with a small set of cases per given dtype generated using `hypothesis`, using the draft code on this gist: https://gist.github.com/tylerjereddy/35936be43743527c16246837295161fe

Is that a reasonable compromise?

File: numpy/core/tests/test_scalar_methods.py (hunk 2 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: this have been parametrized now, so marking as resolved

**Comment by tylerjereddy** [comment]
> it would make sense to me to create a class TestAsIntegerRatio to group them together.

I've done this too now. I suspect Eric or Marten will still have some thoughts on those hard-coded test examples I've added in now though. At first I thought the integer cases didn't seem well-sampled, but I think it is just deceptive because of `minexp` / `maxexp` limits, and the magnitudes on the `ldexp` values produced seem to be on par with the bigger loop originally in place.

**Comment by tylerjereddy** [comment]
Ah, Windows failures on the hardcoded test values. `np.longdouble` looks like the suspect, I think.

File: numpy/core/tests/test_scalar_methods.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                  [0, 1, 0, -8, 12]),
+        (np.single, [0.0, 0.09248576989263226, 0.8160498218131407,
+                     0.17389442853722373, 0.7956044195067877],
+                    [0, 12, 10, 17, -26]),
+        (np.double, [0.0, 0.031066908499895136, 0.5214135908877832,
+                     0.45780736035689296, 0.5906586745934036],
+                    [0, -801, 51, 194, -653]),
+        (np.longdouble, [0.0, 0.20492557202724854, 0.4277180662199366,
+                         0.9888085019891495, 0.9620175814461964],
+                        [0, -7400, 14266, -7822, -8721])])
```
Comment (tylerjereddy) [inline numpy/core/tests/test_scalar_methods.py]: too big for Windows, I think. Not sure if this is worth the effort to break from the approach used upstream really, but could probably just adjust these down a bit I guess

Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: `longdouble` is unsupported in any useful way on windows anyway, so I'm not too worried

File: numpy/core/tests/test_scalar_methods.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    @pytest.mark.parametrize("ftype", [
+        np.half, np.single, np.double, np.longdouble])
+    @pytest.mark.parametrize("f, ratio", [
+        (0.875, (7, 8)),
+        (-0.875, (-7, 8)),
+        (0.0, (0, 1)),
+        (11.5, (23, 2)),
+        ])
+    def test_float_as_integer_ratio(self, ftype, f, ratio):
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: Can you remove the prefix from these names now? Perhaps this could be `test_small`

File: numpy/core/tests/test_scalar_methods.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        assert_equal(R(5, 2),
+                     R(*ftype(2.5).as_integer_ratio()))
+        assert_equal(R(1, 2),
+                     R(*ftype(0.5).as_integer_ratio()))
+        assert_equal(R(-2100, 1),
+                     R(*ftype(-2100.0).as_integer_ratio()))
+
+        assert_raises(OverflowError, ftype('inf').as_integer_ratio)
+        assert_raises(OverflowError, ftype('-inf').as_integer_ratio)
+        assert_raises(ValueError, ftype('nan').as_integer_ratio)
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: These three lines would work well as a `test_errors` test

File: numpy/core/tests/test_scalar_methods.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    def test_float_as_integer_ratio_4(self):
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: `test_against_known_values` perhaps

File: numpy/core/tests/test_scalar_methods.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        (np.single, [0.0, 0.09248576989263226, 0.8160498218131407,
+                     0.17389442853722373, 0.7956044195067877],
+                    [0, 12, 10, 17, -26]),
+        (np.double, [0.0, 0.031066908499895136, 0.5214135908877832,
+                     0.45780736035689296, 0.5906586745934036],
+                    [0, -801, 51, 194, -653]),
+        (np.longdouble, [0.0, 0.20492557202724854, 0.4277180662199366,
+                         0.9888085019891495, 0.9620175814461964],
+                        [0, -7400, 14266, -7822, -8721])])
+    def test_float_as_integer_ratio_2(self, ftype, frac_vals, exp_vals):
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: Since this test is not copied in any form from cpython, let's put it after the ones that are

File: numpy/core/tests/test_scalar_methods.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        rstate = np.random.RandomState(0)
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: Now unused

File: numpy/core/tests/test_scalar_methods.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    @pytest.mark.parametrize("f, ratio", [
+        (0.875, (7, 8)),
+        (-0.875, (-7, 8)),
+        (0.0, (0, 1)),
+        (11.5, (23, 2)),
+        ])
+    def test_float_as_integer_ratio(self, ftype, f, ratio):
+        assert_equal(ftype(f).as_integer_ratio(), ratio)
+
+    @pytest.mark.parametrize("ftype, frac_vals, exp_vals", [
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalar_methods.py]: Now that this is not random, I'm not sure the construction of this test makes much sense.

Perhaps simpler as:
```
@pytest.paramtrize('f', [
    np.half(...),
    np.single(...),
    np.double(...),
    np.longdouble("...")  # note string constructor needed
], id=lambda f: "{}-{}".format(f, f.dtype))
def test_roundtrip(f):
    if f.dtype == np.longdouble and not np.isfinite(f):
        pytest.skip("longdouble too small on this platform")
    n, d = f.as_integer_ratio()
    try:
        nf = np.longdouble(str(n))
        df = np.longdouble(str(d))
    except (OverflowError, RuntimeWarning):
        # the values may not fit in any float type
        continue
    assert nf / df == f
```

File: numpy/core/tests/test_scalar_methods.py (hunk 12 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: removed

File: numpy/core/tests/test_scalar_methods.py (hunk 11 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: ok, done

File: numpy/core/tests/test_scalar_methods.py (hunk 10 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: done

File: numpy/core/tests/test_scalar_methods.py (hunk 9 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: Ok, I made a separate test for that

File: numpy/core/tests/test_scalar_methods.py (hunk 8 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: yup, I used `test_small`

File: numpy/core/tests/test_scalar_methods.py (hunk 7 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: I marked it with an xfail for Windows

File: numpy/core/tests/test_scalar_methods.py (hunk 13 above)
Comment (tylerjereddy) [reply numpy/core/tests/test_scalar_methods.py]: I made all of your recent requested revisions, except for this one, which I only adopted partially. This version isn't any clearer to me, but feel free to just do this if you like.

I'm assuming you're hardcoding the `np.ldexp()` results instead of generating them in a loop.

**Comment by tylerjereddy** [comment]
Well, at least we're seeing something useful from the ppc64le CI here

**Comment by mattip** [comment]
the case that fails: `f = np.ldexp(0.4277180662199366, 14266, dtype='float128')`

File: numpy/core/tests/test_scalar_methods.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+                    [0, -801, 51, 194, -653]),
+        pytest.param(np.longdouble, [0.0, 0.20492557202724854,
+                     0.4277180662199366, 0.9888085019891495,
+                     0.9620175814461964],
+                     [0, -7400, 14266, -7822, -8721],
+                     marks=pytest.mark.xfail(os.name == 'nt',
+                           reason='longdouble too small on Windows'))])
+    def test_roundtrip(self, ftype, frac_vals, exp_vals):
+        for frac, exp in zip(frac_vals, exp_vals):
+            f = np.ldexp(frac, exp, dtype=ftype)
```
Comment (mattip) [inline numpy/core/tests/test_scalar_methods.py]: this fails on ppc64_le for `f = np.ldexp(0.4277180662199366, 14266, dtype='float128')`

Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: Elsewhere I've seen us use the following to work around this:
```python
@pytest.mark.skipif(np.finfo(np.double) == np.finfo(np.longdouble), reason="long double is same as double")
@pytest.mark.skipif(platform.machine().startswith("ppc"), reason="IBM double double")
```

**Comment by eric-wieser** [comment]
Thanks for the test cleanup @tylerjereddy - I should do some more reading on the hypthesis module you're using to help with these.

I amended your last commit to copy the `skipif` style I found elsewhere in umath tests, and fixed some variable names that I screwed up in the original test.

Feel free to tweak them some more, but I think either way this looks good to me.

File: numpy/core/tests/test_scalar_methods.py (hunk 14 above)
Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: CI passing with the above

File: numpy/core/tests/test_scalar_methods.py (hunk 13 above)
Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: > I'm assuming you're hardcoding the np.ldexp() results instead of generating them in a loop.

Yes, exactly. If the values are arbitrary, showing computation steps is maybe not super useful.

> This version isn't any clearer to me

Fine to leave as is then

File: numpy/core/tests/test_scalar_methods.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/tests/test_scalar_methods.py]: Looks fine to me. At this point, this code has been well tested by virtue of going through a bunch of different versions of the test anyway!

**Comment by tylerjereddy** [comment]
Ok, the code changes were already approved by a core dev modulo testing refinements, which are now complete. So, in it goes, thanks Eric

**Comment by eric-wieser** [comment]
Thanks Tyler!
